book-content/chapters/12-the-weird-parts.md

```diff
@@ -0,0 +1,1653 @@
+
+### Solution 4: Iterating over Objects
+
+Let's look at both looping approaches for the `printUser` function.
+
+#### Option 1: Using `Object.keys().forEach()`
+
+The first approach is to use `Object.keys().forEach()` to iterate over the object keys. Inside the `forEach` callback, we'll access the value of the key by using the `key` variable:
+
+```ts twoslash
+// @errors: 7053
+type User = {
+  id: number;
+  name: string;
+};
+// ---cut---
+function printUser(user: User) {
+  Object.keys(user).forEach((key) => {
+    console.log(user[key]);
+  });
+}
+```
+
+This change will have the test case passing, but TypeScript raises a type error on `user[key]`.
+
+The issue is that the `User` interface doesn't have an index signature. In order to get around the type error without modifying the `User` interface, we can use a type assertion on `key` to tell TypeScript that it is of type `keyof User`:
+
+```tsx
+console.log(user[key as keyof User]);
+```
+
+The `keyof User` will be a union of the property names, such as `id` or `name`. And by using `as`, we are telling TypeScript that `key` is a like a more precise string.
+
+With this change, the error goes away - but our code is a little less safe. If our object has an unexpected key, we might get some odd behavior.
+
+#### Option 2: Using a `for` Loop
+
+The `for` loop approach is similar to the `Object.keys().forEach()` approach. We can use a `for` loop and pass in an object instead of a `user`:
+
+```tsx
+function printUser(user: User) {
+  for (const key in user) {
+    console.log(user[key as keyof typeof user]);
+  }
+}
+```
+
+Like before, we need to use `keyof typeof` because of how TypeScript handles excess property checking.
+
+#### Option 3: Widening the Type
+
+Another approach is to widen the type inside the `printUser` function. In this case, we'll specify that the `user` being passed in is a `Record` with a `string` key and an `unknown` value.
+
+In this case, the object being passed in doesn't have to be a `user` since we're just going to be mapping over every key that it receives:
+
+```tsx
+function printUser(obj: Record<string, unknown>) {
+  Object.keys(obj).forEach((key) => {
+    console.log(obj[key]);
+  });
+}
+```
+
+This works on both the runtime and type levels without error.
+
+#### Option 4: `Object.values`
+
+Another way to iterate over the object is to use `Object.values`:
+
+```tsx
+function printUser(user: User) {
+  Object.values(user).forEach(console.log);
+}
+```
+
+This approach avoids the whole issue with the keys, because `Object.values` will return an array of the values of the object. When this option is available, it's a nice way to avoid needing to deal with issue of loosely typed keys.
```

Comment:
There is an interesting observation to be made about `Object.values` here... if `Object.keys` returns `string[]` because the object could have more properties... why the hell this one returns `(string | number)[]`?
```ts
const test2 = Object.values({ a: 1, b: '' })
```

The same could be said about `Object.values`... the object could contain just any properties and just any values. So if TS is concerned about it... the return type here should really be `unknown[]`.